full/src/main/java/apoc/dv/JDBCResource.java

```diff
@@ -0,0 +1,94 @@
+
+    private final String queryParsed;
+
+    public JDBCResource(String name, Map<String, Object> config) {
+        super(Objects.requireNonNull(name, "Field `name` should be defined"),
+                (String) Objects.requireNonNull(config.get("url"), "Field `url` in config should be defined"),
+                (String) Objects.requireNonNull(config.get("desc"), "Field `desc` in config should be defined"),
+                (List<String>) Objects.requireNonNull(config.get("labels"), "Field `labels` in config should be defined"),
+                (String) Objects.requireNonNull(config.get("query"), "Field `query` in config should be defined"),
+                getParameters(config),
+                "JDBC");
+        this.queryParsed = parseQuery(config);
+    }
+
+    private static List<String> getParameters(Map<String, Object> config) {
+        final String query = (String) config.get("query");
+        final long questionMarks = countForQuestionMarks(query);
+        if (questionMarks > 0) {
+            return LongStream.range(0, questionMarks)
+                    .mapToObj(i -> "?")
+                    .collect(Collectors.toList());
+        }
+        return PLACEHOLDER_PATTERN.matcher(query).results()
+                .map(MatchResult::group)
+                .map(String::trim)
+                .filter(StringUtils::isNotBlank)
+                .collect(Collectors.toList());
+    }
+
+    private static String parseQuery(Map<String, Object> config) {
+        String query = (String) config.get("query");
+        return parseQuery(query);
+    }
+
+    private static String parseQuery(String query) {
+        final long questionMarks = countForQuestionMarks(query);
+        if (questionMarks > 0) {
+            return query;
+        }
+        return PLACEHOLDER_PATTERN.matcher(query).replaceAll("?");
+    }
+
+    @Override
+    public long numOfQueryParams() {
+        final long countForQuestionMarks = countForQuestionMarks(query);
+        if (countForQuestionMarks > 0) {
+            return countForQuestionMarks;
+        }
+        return countForQuestionMarks(parseQuery(query));
+    }
+
+    private static long countForQuestionMarks(String query) {
+        if (StringUtils.isBlank(query)) {
+            return 0;
+        }
+        return Stream.of(query.split("[ (),]"))
+                .filter(Objects::nonNull)
+                .map(String::trim)
+                .filter(s -> "?".equals(s))
+                .count();
+    }
+
+    @Override
+    public Pair<String, Map<String, Object>> getProcedureCallWithParams(Object queryParams, Map<String, Object> config) {
+        final long countForQuestionMarks = countForQuestionMarks(query);
+        final List<Object> list;
+        if (countForQuestionMarks > 0) {
+            list = (List<Object>) queryParams;
+        } else {
+            Map<String, Object> queryMap = (Map<String, Object>) queryParams;
+            list = params.stream()
```

Comment:
shouldn't list be empty when there is no question marks, hence no query parameters?

Reply:
I added a validation method for query parameter in order to check, if the number of provided parameters are the same as expected in case of list and if the keys are the same as defined in the VirtualResource

Reply:
ok, I initially misread the code, so it's basically:
```java
if (usingQuestionMarks) {
   // get params from the list - queryParams is a list
} else {
   // using $
   // get params from the map, extract values based on each param name
}
```
?

Reply:
Yes I'm adding the comment into the code